### DISCUSSION TITLE
"incompatible function arguments" when using python bindings on sonoma mac

### DISCUSSION THREAD

**Discussion by samuelpmish** [wjakob/nanobind#515] [Q&A]
Hi, I've got a C++ project with python bindings through nanobind, including a function with signature

```cpp
m.def("func", [](
    nb::ndarray< nb::numpy, const float, nb::shape< nb::any, 8> > foo,
    nb::ndarray< nb::numpy, const float, nb::shape< 2, 3> > bar,
    float baz) { 
  ...
}
```
that I've been successfully using from python with the following pattern
```py
import numpy as np
from my_bindings import func

foo = np.array([
    [-1.5, -0.8, 0,   1.5, -0.8, 0,   0.1, 0.1],
    [-1.5, -0.4, 0,   1.5, -0.4, 0,   0.1, 0.1],
    [-1.5,  0.0, 0,   1.5,  0.0, 0,   0.1, 0.1],
    [-1.5,  0.4, 0,   1.5,  0.4, 0,   0.1, 0.1],
    [-1.5,  0.8, 0,   1.5,  0.8, 0,   0.1, 0.1]
], dtype=float)

bar = np.array([[-1.1, -1.1, -1.1], [1.1, 1.1, 1.1]], dtype=float)

baz = 0.05

(a, b, c) = func(cylinders, bounds, cell_size)
```

Recently, when I updated nanobind I saw that a few things weren't compiling and that I needed to substitute `nb::any` -> `-1`. After making that small change to get things going again,

```cpp
m.def("func", [](                                // VV
    nb::ndarray< nb::numpy, const float, nb::shape< -1, 8> > foo,
    nb::ndarray< nb::numpy, const float, nb::shape< 2, 3> > bar,
    float baz) { 
  ...
}
```
my python tests are complaining that `func` is no longer callable

```py
func(cylinders, bounds, cell_size)

TypeError: func(): incompatible function arguments. The following argument types are supported:
    1. func(arg0: numpy.ndarray[dtype=float32, writable=False, shape=(*, 8)], arg1: numpy.ndarray[dtype=float32, writable=False, shape=(2, 3)], arg2: float, /) -> tuple[numpy.ndarray[dtype=float32, shape=(*, 3)], numpy.ndarray[dtype=uint64, shape=(*, 3)], numpy.ndarray[dtype=uint64, shape=(*, 4)]]
```

I looked through the documentation to see if there was anything else I might be missing, and there's this note:
https://nanobind.readthedocs.io/en/latest/ndarray.html#why-does-nanobind-not-accept-my-numpy-array

So, I tried removing the `const`ness in the C++ code, and (separately) doing `setflags(write=false)` on the numpy arrays and was still unable to call this function (same `TypeError` message)

Can anyone help me understand what I'm doing wrong with the recent version of nanobind?

------------

Edit: after reverting to an older version of nanobind, I'm getting a similar error. It occurs to me, this is also my first time building this project on a new mac (w/ python 3.12 + numpy 1.26.4).

**Comment by hpkfft** [comment]
In C++, `float` is IEEE single precision:
```
nb::ndarray< nb::numpy, const float, nb::shape< 2, 3> > bar
```
I believe that in python, `float` is double precision:
```
bar = np.array([[-1.1, -1.1, -1.1], [1.1, 1.1, 1.1]], dtype=float)
```
I believe you may have intended `dtype=np.float32` in the line immediately above.

By the way, an `nb::ndarray` using const-qualified scalar type can bind to python arrays that are writable or not.  It indicates that the C++ code will not be writing the array, so the array is allowed to be read-only.   There is no reason to `setflags(write=false)` unless you want that for some reason in your python code.

**Reply by samuelpmish** [reply]
Thank you for the suggestion, but replacing `dtype=float` by `dtype=np.float32` or `dtype=np.single` appears to yield the same error message.

https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.single

**Reply by samuelpmish** [reply]
To make matters more confusing, the same C++ project w/ bindings and same test run without issue on a linux machine with python 3.10.12

**Comment by hpkfft** [comment]
Hmmm, I'm puzzled why this is not working.
Just to be thorough, maybe it's worth trying `double baz` in your binding (assuming `cell_size` is a python float).
Or maybe `baz` needs to be an `ndarray` (like `foo` and `bar`) so it can have a custom dtype....

Sorry, I'm out of ideas and I have to run.   Anyway, my experience is basically only Linux.
